src/Functions/array/array.cpp

```diff
@@ -77,35 +82,198 @@ class FunctionArray : public IFunction
+                out_data.insertFrom(*column, i);
+        return true;
+    }
+
+    template <typename T>
+    bool executeNumber(const ColumnRawPtrs & columns, IColumn & out_data, size_t input_rows_count) const
+    {
+        using Container = ColumnVectorOrDecimal<T>::Container;
+        std::vector<const Container *> containers(columns.size(), nullptr);
+        for (size_t i = 0; i < columns.size(); ++i)
+        {
+            const ColumnVectorOrDecimal<T> * concrete_column = checkAndGetColumn<ColumnVectorOrDecimal<T>>(columns[i]);
+            if (!concrete_column)
+                return false;
+
+            containers[i] = &concrete_column->getData();
+        }
+
+        ColumnVectorOrDecimal<T> & concrete_out_data = assert_cast<ColumnVectorOrDecimal<T> &>(out_data);
+        Container & out_container = concrete_out_data.getData();
+        out_container.resize_exact(columns.size() * input_rows_count);
+
+        for (size_t row_i = 0; row_i < input_rows_count; ++row_i)
+        {
+            const size_t base = row_i * columns.size();
+            for (size_t col_i = 0; col_i < columns.size(); ++col_i)
+                out_container[base + col_i] = (*containers[col_i])[row_i];
+        }
+        return true;
+    }
+
+    bool executeString(const ColumnRawPtrs & columns, IColumn & out_data, size_t input_rows_count) const
+    {
+        size_t total_bytes = 0;
+        std::vector<const ColumnString *> concrete_columns(columns.size(), nullptr);
+        for (size_t i = 0; i < columns.size(); ++i)
+        {
+            const ColumnString * concrete_column = checkAndGetColumn<ColumnString>(columns[i]);
+            if (!concrete_column)
+                return false;
+
+            total_bytes += concrete_column->getChars().size();
+            concrete_columns[i] = concrete_column;
+        }
+
+        ColumnString & concreate_out_data = assert_cast<ColumnString &>(out_data);
+        auto & out_chars = concreate_out_data.getChars();
+        auto & out_offsets = concreate_out_data.getOffsets();
+        out_chars.resize_exact(total_bytes);
+        out_offsets.resize_exact(input_rows_count * columns.size());
+
+        size_t curr_out_offset = 0;
+        for (size_t row_i = 0; row_i < input_rows_count; ++row_i)
+        {
+            const size_t base = row_i * columns.size();
+            for (size_t col_i = 0; col_i < columns.size(); ++col_i)
+            {
+                StringRef ref = concrete_columns[col_i]->getDataAt(row_i);
+                memcpySmallAllowReadWriteOverflow15(&out_chars[curr_out_offset], ref.data, ref.size);
+                out_chars[curr_out_offset + ref.size] = 0;
+
+                curr_out_offset += ref.size + 1;
+                out_offsets[base + col_i] = curr_out_offset;
+            }
+        }
+        return true;
+    }
+
+    bool executeFixedString(const ColumnRawPtrs & columns, IColumn & out_data, size_t input_rows_count) const
+    {
```

Comment:
There're many common codes between the `executeString` and the `executeFixedString` function.

Would it better to extract them into a function named `executeStringOrFixedString` firstly?

Reply:
I thought about it before and I decided to use isolated functions for `ColumnString` and `ColumnFixedString`. Because each value in ColumnString has a suffix '\0', even if we have a function `executeStringOrFixedString`, we still need `std::is_same_v(T, ColumnString)`, which is not graceful enough.

Reply:
Actually, when many functions' code handle with **string** and **fixedstring**, they put them together, like:
`executeString()` in the https://github.com/ClickHouse/ClickHouse/blob/master/src/Functions/FunctionsComparison.h

So that I suggest the same way here. 

@rschu1ze May I ask what's your opinion?

Reply:
Everyone is currently busy fixing CI, let's postpone this discussion a bit.